docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+
+The below diagram shows the detailed inheritance hierarchy for `UnbufferedSerial`.
+
+![UnbufferedSerial class](./diagrams/unbuffered-serial.png)
+
+**API description**
+
+This is the proposed API for the `UnbufferedSerial` class.
+
+```C
+class UnbufferedSerial: public SerialBase, public FileHandle, private NonCopyable<UnbufferedSerial> {
+
+public:
+    /** Create a UnbufferedSerial port, connected to the specified transmit and receive pins, with the specified baud.
+     *
+     *  @param tx Transmit pin
+     *  @param rx Receive pin
+     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
+     *
+     *  @note
+     *    Either tx or rx may be specified as NC if unused
+     */
+    UnbufferedSerial(PinName tx, PinName rx, int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);
+
+    /** Write a char to the serial port
+     *
+     * @param c The char to write
+     *
+     * @returns The written char or -1 if an error occurred
+     */
+    int putc(int c);
+
+    /** Read a char from the serial port
+     *
+     * @returns The char read from the serial port
+     */
+    int getc();
+
+    /** Write a string to the serial port
+     *
+     * @param str The string to write
+     *
+     * @returns 0 if the write succeeds, EOF for error
+     */
+    int puts(const char *str);
+
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET);
+    virtual off_t size();
+    virtual int isatty();
+    virtual int close();
+    virtual short poll(short events) const;
+
+#if !(DOXYGEN_ONLY)
+protected:
+
+    /* Acquire exclusive access to this serial port
+     */
+    virtual void lock(void);
+
+    /* Release exclusive access to this serial port
+     */
+    virtual void unlock(void);
+#endif
+};
+```
+
+###  Detailed design : MinimalSerial
+
+The `MinimalSerial` class is a minimal abstract base class that provides basics I/O functionality. All other serial classes derive from it.
+
+**API description**
+
+This is the proposed API for the `MinimalSerial` class. Only two methods are provided to output `_base_putc()` and receive `_base_getc()` characters on the UART.
+
+```C
+class MinimalSerial : private NonCopyable<MinimalSerial> {
+protected:
+    MinimalSerial(PinName tx, PinName rx, int baud);
+    virtual ~MinimalSerial();
+    int _base_getc();
+    int _base_putc(int c);
+};
+```
+
+###  Detailed design : MinimalConsole
+
+`MinimalConsole` is an internal concrete class that derives from `MinimalSerial` and provides basic console functionality that can be used by the retarget code for resources constrained targets, see [Simplication of retarget code](#simplification-of-retarget-code).
```

Comment:
I'm not convinced this wants to be tied in with any of the `FileHandle` related stuff. The extra virtual overhead of two more classes doesn't seem to be buying anything. And in the simplified retarget, nothing needs/wants to be virtual to keep the size down.

And `MinimalConsole` shouldn't be locked into serial anyway, in which case having both `SerialBase` and `MinimalConsole` based on `MinimalSerial` isn't logical.

If offering the minimal console, it should be based on weak functions to allow the possibility to override (even though it's a parallel override mechanism to the FileHandle-based one, there's no reason not to do it, and plenty of similar weak precedent).

Reply:
> The extra virtual overhead of two more classes doesn't seem to be buying anything. And in the simplified retarget, nothing needs/wants to be virtual to keep the size down.

I think you might be right. If it increases the simplified retarget then it will be removed. I don't have a strong opinion on `MinimalSerial` other than the fact that it provides basic functionality for both a console and `SerialBase`.

Reply:
MinimalConsole was removed

Reply:
> MinimalConsole was removed

Was it? when?